.tools/create_algo_selection_code.py

```diff
@@ -0,0 +1,415 @@
+import importlib
+import inspect
+import pkgutil
+import textwrap
+from itertools import combinations
+from types import ModuleType
+from typing import Callable, Type
+
+from optimagic.config import OPTIMAGIC_ROOT
+from optimagic.optimization.algorithm import Algorithm
+from optimagic.typing import AggregationLevel
+
+
+def main():
+    """Create the source code for algorithms.py.
+
+    The main part are nested dataclasses that enable filtered autocomplete for algorithm
+    selection.
+
+    """
+    # create some basic inputs
+    docstring = _get_docstring_code()
+    modules = _import_optimizer_modules("optimagic.optimizers")
+    all_algos = _get_all_algorithms(modules)
+    filters = _get_filters()
+    all_categories = list(filters)
+    selection_info = _create_selection_info(all_algos, all_categories)
+
+    # create the code for imports
+    imports = _get_imports(modules)
+
+    # create the code for the ABC AlgoSelection class
+    parent_class_snippet = _get_base_class_code()
+
+    # create the code for the dataclasses
+    dataclass_snippets = []
+    for active_categories in selection_info:
+        new_snippet = create_dataclass_code(
+            active_categories=active_categories,
+            all_categories=all_categories,
+            selection_info=selection_info,
+        )
+        dataclass_snippets.append(new_snippet)
+
+    # create the code for the instantiation
+    instantiation_snippet = _get_instantiation_code()
+
+    # write code to the file
+
+    with open(OPTIMAGIC_ROOT / "algorithms.py", "w") as f:
+        f.write(docstring)
+        f.write(imports + "\n\n")
+        f.write(parent_class_snippet + "\n")
+        f.write("\n\n".join(dataclass_snippets))
+        f.write("\n\n")
+        f.write(instantiation_snippet)
```

Comment:
Any reason not to combine the code into a single string beforehand, and use pathlib.Path's write functionality?

```python
# Combine all the content into a single string
content = (
    docstring +
    imports + "\n\n" +
    parent_class_snippet + "\n" +
    "\n\n".join(dataclass_snippets) + "\n\n" +
    instantiation_snippet
)

# Write the combined content to the file
with (OPTIMAGIC_ROOT / "algorithms.py").open("w") as f:
    f.write(content)
```